### PR TITLE
crimson/os/seastore/btree: clean up `FixedKVLeafNode::get_logical_child()`

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/crimson/os/seastore/btree/fixed_kv_node.h [hunk 1]
```diff
@@ -972,27 +970,7 @@ struct FixedKVLeafNode
...
-	ceph_assert(child->is_logical());
-	return c.cache.template get_extent_viewable_by_trans<
-	  LogicalCachedExtent>(c.trans, (LogicalCachedExtent*)child);
-      } else {
-	return child_pos_t(&sparent, spos);
-      }
-    } else {
-      return child_pos_t(this, pos);
-    }
+    return this->template get_child<LogicalCachedExtent>(c, this->iter_idx(pos));
```
Comment (cyx1231st) [inline src/crimson/os/seastore/btree/fixed_kv_node.h:973]: Looking at the user of `get_logical_child()` -- seems `BtreeNodeMapping` already has both `pos` and `key`.

Does it mean `BtreeNodeMapping` can directly call `get_child()`, and `get_logical_child()` can be dropped ?

File: src/crimson/os/seastore/btree/btree_range_pin.cc [hunk 2]
```diff
@@ -15,7 +15,7 @@ BtreeNodeMapping<key_t, val_t>::get_logical_extent(
   assert(parent->is_valid());
   assert(pos != std::numeric_limits<uint16_t>::max());
   auto &p = (FixedKVNode<key_t>&)*parent;
-  auto v = p.get_logical_child(ctx, pos);
+  auto v = p.template get_child<LogicalCachedExtent>(ctx, pos, get_key());
```
Comment (liu-chunmei) [inline src/crimson/os/seastore/btree/btree_range_pin.cc:18]: can we assert here parent should be FixedKVLeafNode?

File: src/crimson/os/seastore/btree/fixed_kv_node.h (hunk 1 above)
Comment (liu-chunmei) [reply src/crimson/os/seastore/btree/fixed_kv_node.h]: I think so, get_child() include get_logical_child().

**Comment by cyx1231st** [comment]
Looks good, but somehow triggers the ceph-abort from make check:

ERROR 2024-04-02 09:47:38,607 [shard 0:main] none - /home/jenkins-build/build/workspace/ceph-pull-requests/src/crimson/os/seastore/cache.h:413 : In function 'get_extent_iertr::future<TCachedExtentRef<T>> crimson::os::seastore::Cache::get_absent_extent(crimson::os::seastore::Transaction &, crimson::os::seastore::paddr_t, crimson::os::seastore::extent_len_t, Func &&) [T = crimson::os::seastore::TestBlock, Func = (lambda at /home/jenkins-build/build/workspace/ceph-pull-requests/src/crimson/os/seastore/transaction_manager.h:860:7)]', abort(%s)

File: src/crimson/os/seastore/btree/btree_range_pin.cc (hunk 2 above)
Comment (xxhdx1985126) [reply src/crimson/os/seastore/btree/btree_range_pin.cc:18]: will try.

**Comment by xxhdx1985126** [comment]
> Looks good, but somehow triggers the ceph-abort from make check:
> 
> ERROR 2024-04-02 09:47:38,607 [shard 0:main] none - /home/jenkins-build/build/workspace/ceph-pull-requests/src/crimson/os/seastore/cache.h:413 : In function 'get_extent_iertr::future<TCachedExtentRef> crimson::os::seastore::Cache::get_absent_extent(crimson::os::seastore::Transaction &, crimson::os::seastore::paddr_t, crimson::os::seastore::extent_len_t, Func &&) [T = crimson::os::seastore::TestBlock, Func = (lambda at /home/jenkins-build/build/workspace/ceph-pull-requests/src/crimson/os/seastore/transaction_manager.h:860:7)]', abort(%s)

Looking into it

**Comment by xxhdx1985126** [comment]
@cyx1231st Fixed, please take a look again.

**Comment by xxhdx1985126** [comment]
jenkins test windows

**Review by cyx1231st — APPROVED** [review]

**Comment by xxhdx1985126** [comment]
jenkins test make check

**Comment by xxhdx1985126** [comment]
> The following tests FAILED:
	 22 - run-tox-cephadm (Failed)

**Comment by xxhdx1985126** [comment]
jenkins test make check

**Comment by xxhdx1985126** [comment]
jenkins test api

**Pull request merged**